pkg/markers/collect.go

```diff
@@ -31,7 +31,7 @@ import (
 type Collector struct {
 	*Registry
 
-	byPackage map[string]map[ast.Node]MarkerValues
+	byPackage map[*loader.Package]map[ast.Node]MarkerValues
```

Comment:
This stores the package based on the pointer's value, is there any other way that doesn't rely on where the memory is allocated?

Reply:
There is a unique ID in https://pkg.go.dev/golang.org/x/tools/go/packages#Package

Reply:
Thank you for your reviewing!

There is possibly another way, but it may be complicated.

`byPackage` has two keys, package ID (`string`) and node (`ast.Node`). Package ID is not a pointer's value, while node is an interface and can be a struct containing pointers. That inconsistency is the root cause of this issue.

In the problematic situation, exactly the same package is loaded several times. Among each load, the package ID is the same but nodes are different objects. [Parsing markers is skipped](https://github.com/kubernetes-sigs/controller-tools/blob/73391d1c6f3edfed27bbd82c8cc7bfb05f635bad/pkg/markers/collect.go#L78) when the same package ID is already processed. Therefore, `byPackage`'s node keys are created only at the first time. As a result, access `byPackage` using node as a key fails from the second time.

To avoid to use pointer's value as `byPackage`'s key, I think we have to do one of the following.

* Stop using `ast.Node` as a key (but I have no idea about the alternative)
* Do not skip parsing markers even if a package with the same ID is already parsed
* Stop loading the same package multiple times (but I have no idea why the multiple load occurs)

Reply:
Yup agree. The problem is definitely that we cache the markers (with pointers) by package.ID and then try to use them for another package (which has different ast nodes).

Reply:
I think it should be not worse to use *loader.Package as a map key then it already is to use the ast.Node interface (aka e.g. *ast.File) as a map key one layer deeper)

Reply:
That's fair, it's a bit triggering though in review 😄

Reply:
Yeah 100% :)